src/DacpacTool/ScriptParser.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Linq;
+using System.Reflection;
+using Microsoft.SqlTools.ServiceLayer.BatchParser;
+
+namespace MSBuild.Sdk.SqlProj.DacpacTool
+{
+    public class ScriptParser : ICommandHandler
+    {
+        private readonly Parser _parser;
+        private bool _parsed;
+        private readonly List<string> _includedFileNames = new List<string>();
+        private StringBuilder _scriptBuilder = new StringBuilder();
+        private string _currentFilePath;
+
+        public ScriptParser(string sourceFile, IVariableResolver variableResolver)
+        {
+            if (Path.IsPathRooted(sourceFile))
+            {
+                _currentFilePath = Path.GetDirectoryName(sourceFile);
+            }
+            else
+            {
+                sourceFile = Path.Combine(Environment.CurrentDirectory, sourceFile);
+                _currentFilePath = Path.GetDirectoryName(sourceFile);
+            }
+
+            _parser = new Parser(this, variableResolver, new StreamReader(sourceFile), sourceFile);
+        }
+
+        protected void Parse()
+        {
+            if (! _parsed)
+            {
+                _parser.Parse();
+                _parsed = true;
+            }
+        }
+
+        public List<string> CollectFileNames()
+        {
+            Parse();
+            return _includedFileNames;
+        }
+
+        public string GenerateScript()
+        {
+            Parse();
+            return _scriptBuilder.ToString();
+        }
+
+        BatchParserAction ICommandHandler.Go(TextBlock batch, int repeatCount, SqlCmdCommand tokenType)
+        {
+            batch.GetText(true, out string batchText, out LineInfo lineInfo);
+            _scriptBuilder.Append(batchText);
+            _scriptBuilder.Append("\nGO\n");
+            return BatchParserAction.Continue;
+        }
+
+        BatchParserAction ICommandHandler.Include(TextBlock includeBlock, out TextReader stream, out string newFilename)
+        {
+            includeBlock.GetText(true, out string includedFileName, out LineInfo lineInfo);
+
+            // HACK: Work around access limitations to get access to the underlying tokens
+            var tokens = 
+                includeBlock.GetType()
+                            .GetField("tokens", BindingFlags.NonPublic | BindingFlags.Instance)
+                            .GetValue(includeBlock) as IEnumerable<Token>;
+            var token = tokens.FirstOrDefault();
+            if (token != null)
+            {
+                _currentFilePath = Path.GetDirectoryName(token.Filename);
+            }
```

Comment:
Why do we need access to the tokens? Isn't everything we need here contained in `includedFileName`?

Reply:
The tokens are the only place where we can get access to the path information tracked by `Microsoft.SqlTools.ServiceLayer.BatchParser.Lexer.inputStack`. The problem is that, within recursively-included files, each included filename is relative to the current path of the file it's included within.

This was the only way I could resolve the problem posed by setups like `test/TestProjectWithPrePost/Pre-Deployment/Script.PreDeployment.NestedRelativePaths.sql` -- which is based on a real-life issue I had when trying to build one of my own projects.

Reply:
I did some digging and it turns out that through the `LineInfo` object returned by `GetText()` on the `includeBlock` has a method that we can use to get the original position from which the include was being made including the filename. I'm a little bit unsure about the argument you need to pass into that function though. I'm currently just passing in `0` which seems to work, or at least it passes the tests.

Reply:
Thanks, wow, it never occurred to me at all that `Filename` would be a member of something called `PositionStruct`, thanks for finding that!

Reply:
Yeah, it doesn't make much sense to me either. It seems that the `GetStreamPositionForOffset()` method I'm using is also called internally with a hardcoded argument of `0` so I think we can accept this as a viable option.

Reply:
Agreed, I'm just going to test it quickly with my real-life project to confirm

Reply:
Of course! Let me know how that goes ;).

Reply:
Confirmed that this works for me, and passes all tests. Looks so much better!